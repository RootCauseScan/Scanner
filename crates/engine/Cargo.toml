[package]
name = "engine"
version = "1.0.0"
edition = "2021"
description = "Parallel rule evaluation engine for RootCause"
readme = "README.md"
license.workspace = true
repository = "https://github.com/rootcausescan/scanner"
homepage = "https://github.com/rootcausescan/scanner"
keywords = ["security", "sast", "static-analysis"]
categories = ["development-tools", "code-analysis"]

[dependencies]
rayon = { workspace = true }
regex = { workspace = true }
serde = { workspace = true, features = ["derive"] }
serde_json = { workspace = true }
loader = { path = "../loader" }
ir = { path = "../ir" }
rego_wasm = { path = "../rego_wasm" }
tokio = { version = "1", features = ["rt-multi-thread", "time"] }
anyhow = "1"
tracing = { workspace = true }
blake3 = "1"
parsers = { path = "../parsers" }
shlex = { workspace = true }
plugin_core = { path = "../plugin_core" }

[dev-dependencies]
parsers = { path = "../parsers" }
tempfile = "3"
tracing-subscriber = { workspace = true }
base64 = { workspace = true }
