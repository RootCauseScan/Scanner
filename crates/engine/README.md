# engine

Analysis engine that evaluates rules over the intermediate representation with parallel execution and cache.

## Build

```bash
cargo build -p engine
```

## Test

```bash
cargo test -p engine
```

## Documentation

- [Architecture details](../../documentation/architecture/crates/engine.md)

## Contributing

See [CONTRIBUTING.md](../CONTRIBUTING.md) for guidelines.
