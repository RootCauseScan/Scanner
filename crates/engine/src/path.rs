use regex::Regex;
use std::collections::{HashMap, VecDeque};
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::{Mutex, OnceLock, RwLock};

pub static CANONICAL_PATHS: OnceLock<RwLock<HashMap<PathBuf, PathBuf>>> = OnceLock::new();
pub static CANONICAL_STATS: OnceLock<CacheStats> = OnceLock::new();
pub static CANONICAL_ORDER: OnceLock<RwLock<VecDeque<PathBuf>>> = OnceLock::new();
static PATH_REGEX_CACHE: OnceLock<Mutex<PathRegexCache>> = OnceLock::new();

#[cfg(test)]
pub const CANONICAL_CACHE_CAPACITY: usize = 3;
#[cfg(not(test))]
pub const CANONICAL_CACHE_CAPACITY: usize = 1024;

#[cfg(test)]
pub const PATH_REGEX_CACHE_CAPACITY: usize = 3;
#[cfg(not(test))]
pub const PATH_REGEX_CACHE_CAPACITY: usize = 1024;

#[derive(Default)]
pub struct CacheStats {
    hits: AtomicUsize,
    misses: AtomicUsize,
}

struct PathRegexCache {
    map: HashMap<String, Regex>,
    order: VecDeque<String>,
    capacity: usize,
}

impl PathRegexCache {
    fn new(capacity: usize) -> Self {
        Self {
            map: HashMap::new(),
            order: VecDeque::new(),
            capacity,
        }
    }

    fn get(&mut self, key: &str) -> Option<&Regex> {
        if let Some(pos) = self.order.iter().position(|k| k == key) {
            let k = self.order.remove(pos).expect("order index must exist");
            self.order.push_back(k);
            return self.map.get(key);
        }
        None
    }

    fn insert(&mut self, key: String, value: Regex) {
        if self.map.contains_key(&key) {
            if let Some(pos) = self.order.iter().position(|k| k == &key) {
                self.order.remove(pos);
            }
        }
        self.order.push_back(key.clone());
        self.map.insert(key, value);
        if self.order.len() > self.capacity {
            if let Some(oldest) = self.order.pop_front() {
                self.map.remove(&oldest);
            }
        }
    }

    #[cfg(test)]
    fn clear(&mut self) {
        self.map.clear();
        self.order.clear();
    }

    #[cfg(test)]
    fn len(&self) -> usize {
        self.map.len()
    }

    #[cfg(test)]
    fn contains(&self, key: &str) -> bool {
        self.map.contains_key(key)
    }
}

pub fn canonicalize_path<P: AsRef<Path>>(path: P) -> PathBuf {
    let path_ref = path.as_ref();
    let cache = CANONICAL_PATHS.get_or_init(|| RwLock::new(HashMap::new()));
    let order = CANONICAL_ORDER.get_or_init(|| RwLock::new(VecDeque::new()));
    let stats = CANONICAL_STATS.get_or_init(Default::default);
    if let Some(cached) = cache
        .read()
        .unwrap_or_else(|e| e.into_inner())
        .get(path_ref)
        .cloned()
    {
        stats.hits.fetch_add(1, Ordering::Relaxed);
        let mut ord = order.write().unwrap_or_else(|e| e.into_inner());
        if let Some(pos) = ord.iter().position(|p| p == path_ref) {
            ord.remove(pos);
        }
        ord.push_back(path_ref.to_path_buf());
        return cached;
    }
    stats.misses.fetch_add(1, Ordering::Relaxed);
    let canonical = fs::canonicalize(path_ref).unwrap_or_else(|_| {
        #[cfg(windows)]
        {
            PathBuf::from(path_ref.to_string_lossy().replace('\\', "/"))
        }
        #[cfg(not(windows))]
        {
            path_ref.to_path_buf()
        }
    });
    let mut map = cache.write().unwrap_or_else(|e| e.into_inner());
    let mut ord = order.write().unwrap_or_else(|e| e.into_inner());
    map.insert(path_ref.to_path_buf(), canonical.clone());
    ord.push_back(path_ref.to_path_buf());
    if ord.len() > CANONICAL_CACHE_CAPACITY {
        if let Some(oldest) = ord.pop_front() {
            map.remove(&oldest);
        }
    }
    canonical
}

pub(crate) fn cache_stats() -> (usize, usize) {
    let stats = CANONICAL_STATS.get_or_init(Default::default);
    (
        stats.hits.load(Ordering::Relaxed),
        stats.misses.load(Ordering::Relaxed),
    )
}

#[cfg(test)]
pub fn reset_canonical_cache() {
    if let Some(map) = CANONICAL_PATHS.get() {
        map.write().unwrap_or_else(|e| e.into_inner()).clear();
    }
    if let Some(ord) = CANONICAL_ORDER.get() {
        ord.write().unwrap_or_else(|e| e.into_inner()).clear();
    }
    if let Some(stats) = CANONICAL_STATS.get() {
        stats.hits.store(0, Ordering::Relaxed);
        stats.misses.store(0, Ordering::Relaxed);
    }
}

#[cfg(test)]
pub fn canonical_cache_stats() -> (usize, usize) {
    cache_stats()
}

#[cfg(test)]
pub fn reset_path_regex_cache() {
    if let Some(cache) = PATH_REGEX_CACHE.get() {
        cache.lock().unwrap().clear();
    }
}

#[cfg(test)]
pub fn path_regex_cache_size() -> usize {
    PATH_REGEX_CACHE
        .get()
        .map(|c| c.lock().unwrap().len())
        .unwrap_or(0)
}

#[cfg(test)]
pub fn path_regex_cache_contains(pat: &str) -> bool {
    PATH_REGEX_CACHE
        .get()
        .map(|c| c.lock().unwrap().contains(pat))
        .unwrap_or(false)
}

pub fn path_matches(pattern: &str, candidate: &str) -> bool {
    if !pattern
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || "/._-".contains(c) || c == '*')
    {
        return false;
    }
    let cache =
        PATH_REGEX_CACHE.get_or_init(|| Mutex::new(PathRegexCache::new(PATH_REGEX_CACHE_CAPACITY)));
    let mut cache = cache.lock().expect("path regex cache lock poisoned");
    if let Some(rx) = cache.get(pattern) {
        return rx.is_match(candidate);
    }
    let mut re = String::from("^");
    for ch in pattern.chars() {
        if ch == '*' {
            re.push_str(".*");
        } else {
            re.push_str(&regex::escape(&ch.to_string()));
        }
    }
    re.push('$');
    match Regex::new(&re) {
        Ok(rx) => {
            let is_match = rx.is_match(candidate);
            cache.insert(pattern.to_string(), rx);
            is_match
        }
        Err(_) => false,
    }
}
