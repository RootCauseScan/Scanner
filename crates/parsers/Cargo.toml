[package]
name = "parsers"
version = "1.0.0"
edition = "2021"
description = "Parsers that convert source files into RootCause IR"
readme = "README.md"
license.workspace = true
repository = "https://github.com/rootcausescan/scanner"
homepage = "https://github.com/rootcausescan/scanner"
keywords = ["security", "sast", "static-analysis"]
categories = ["development-tools", "code-analysis"]

[dependencies]
anyhow = "1"
rayon = { workspace = true }
regex = { workspace = true }
serde = { workspace = true,  features = ["derive"] }
serde_yaml = { workspace = true }
serde_json = { workspace = true }
ir = { path = "../ir" }
tree-sitter = "0.20"
tree-sitter-typescript = "0.20"
tree-sitter-python = "0.20"
tree-sitter-go = "0.20"
tree-sitter-ruby = "0.20"
tree-sitter-rust = "0.20"
tree-sitter-javascript = "0.20"
tree-sitter-java = "0.20"
tree-sitter-php = "0.20"
marked-yaml = "0.8"
hcl-edit = "0.9"
tracing = { workspace = true }
blake3 = "1"

[dev-dependencies]
tempfile = "3"

