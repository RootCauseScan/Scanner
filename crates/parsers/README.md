# parsers

Collection of parsers for Dockerfile, YAML, HCL and various code languages. Generates `FileIR` and, optionally, ASTs.

## Build

```bash
cargo build -p parsers
```

## Test

```bash
cargo test -p parsers
```

## Documentation

- [Supported languages](../../documentation/architecture/crates/parsers)

## Contributing

See [CONTRIBUTING.md](../CONTRIBUTING.md) for guidelines.
