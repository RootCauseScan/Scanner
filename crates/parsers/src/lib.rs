//! Converters of configuration and code files to the
//! intermediate representation used by the engine.

use anyhow::{anyhow, Context, Result};
use ir::FileIR;
use serde::{Deserialize, Serialize};
use std::{
    collections::{HashMap, HashSet},
    fs,
    path::Path,
};
use tracing::debug;

#[derive(Debug, Default, Serialize)]
pub struct ParserMetrics {
    pub files_parsed: usize,
    pub cache_hits: usize,
    pub parse_errors: usize,
}

pub mod catalog;
pub mod languages;
pub use languages::*;

/// Builds the data flow graph if it doesn't already exist.
/// For supported languages, it re-parses the code using
/// the corresponding parser.
pub fn build_dfg(fir: &mut FileIR) -> Result<()> {
    if fir.dfg.is_some() {
        return Ok(());
    }
    let src = fir
        .source
        .clone()
        .ok_or_else(|| anyhow!("missing source for {}", fir.file_path))?;
    match fir.file_type.as_str() {
        "python" => languages::python::parse_python(&src, fir),
        "rust" => languages::rust::parse_rust(&src, fir),
        _ => Ok(()),
    }
}

/// Determines the supported file type from the name/extension.
///
/// # Example
/// ```
/// use parsers::detect_type;
/// assert_eq!(detect_type(std::path::Path::new("Dockerfile")), Some("dockerfile"));
/// ```
pub fn detect_type(path: &Path) -> Option<&'static str> {
    let name = path.file_name()?.to_string_lossy().to_lowercase();
    let ext = path.extension().map(|e| e.to_string_lossy().to_lowercase());
    let detected = if name == "dockerfile" {
        Some("dockerfile")
    } else {
        match ext.as_deref() {
            Some("yaml") | Some("yml") => Some("yaml"),
            Some("json") => Some("json"),
            Some("tf") => Some("hcl"),
            Some("ts") | Some("tsx") => Some("typescript"),
            Some("js") | Some("jsx") => Some("javascript"),
            Some("py") => Some("python"),
            Some("go") => Some("go"),
            Some("rb") => Some("ruby"),
            Some("rs") => Some("rust"),
            Some("java") => Some("java"),
            Some("php") => Some("php"),
            _ => Some("generic"),
        }
    };
    if let Some(t) = detected {
        debug!(file = %path.display(), file_type = t, "File type detected");
    } else {
        debug!(file = %path.display(), "Unsupported file type");
    }
    detected
}

/// Analiza contenido YAML desde una cadena y genera un [`FileIR`].
///
/// # Example
/// ```
/// use parsers::parse_str;
/// let ir = parse_str("a: 1").unwrap();
/// assert_eq!(ir.nodes[0].path, "a");
/// ```
pub fn parse_str(content: &str) -> anyhow::Result<FileIR> {
    let mut fir = FileIR::new("<memory>".to_string(), "yaml".to_string());
    fir.source = Some(content.to_string());
    parse_yaml(content, &mut fir)?;
    Ok(fir)
}

/// Reads a file and produces its intermediate representation [`FileIR`].
///
/// # Example
/// ```
/// use parsers::parse_file;
/// use std::fs;
/// let path = std::env::temp_dir().join("ex.yaml");
/// fs::write(&path, "a: 1\n# note").unwrap();
/// let ir = parse_file(&path, None, None).unwrap().unwrap();
/// assert_eq!(ir.nodes[0].path, "a");
/// assert!(ir.source.is_some());
/// ```
pub fn parse_file(
    path: &Path,
    suppress_comment: Option<&str>,
    mut metrics: Option<&mut ParserMetrics>,
) -> anyhow::Result<Option<FileIR>> {
    debug!(file = %path.display(), "Detecting file type");
    let Some(ftype) = detect_type(path) else {
        return Ok(None);
    };
    debug!(file = %path.display(), file_type = ftype, "Parsing file");
    let content = fs::read_to_string(path)
        .with_context(|| format!("Failed to read file: {}", path.display()))?;
    let suppressed = suppress_comment.map_or_else(HashSet::new, |c| {
        content
            .lines()
            .enumerate()
            .filter_map(|(idx, line)| {
                if line.contains(c) {
                    Some(idx + 1)
                } else {
                    None
                }
            })
            .collect()
    });
    let canonical = fs::canonicalize(path).unwrap_or_else(|_| path.to_path_buf());
    let mut fir = FileIR::new(canonical.to_string_lossy().into_owned(), ftype.to_string());
    let res: anyhow::Result<()> = match ftype {
        "dockerfile" => {
            parse_dockerfile(&content, &mut fir);
            Ok(())
        }
        "yaml" => parse_yaml(&content, &mut fir),
        "json" => parse_json(&content, &mut fir),
        "hcl" => {
            parse_hcl(&content, &mut fir);
            Ok(())
        }
        "typescript" => {
            parse_typescript(&content, &mut fir);
            Ok(())
        }
        "javascript" => {
            parse_javascript(&content, &mut fir);
            Ok(())
        }
        "python" => languages::python::parse_python(&content, &mut fir),
        "go" => {
            parse_go(&content, &mut fir);
            Ok(())
        }
        "ruby" => {
            parse_ruby(&content, &mut fir);
            Ok(())
        }
        "rust" => parse_rust(&content, &mut fir),
        "java" => parse_java(&content, &mut fir),
        "php" => parse_php(&content, &mut fir),
        "generic" => parse_generic(&content, &mut fir),
        _ => Ok(()),
    };
    if let Err(e) = res {
        if let Some(m) = metrics.as_deref_mut() {
            m.parse_errors += 1;
        }
        return Err(e);
    }
    if fir.symbol_types.contains_key("__parse_error__") {
        if let Some(m) = metrics.as_deref_mut() {
            m.parse_errors += 1;
        }
        return Ok(None);
    }
    if let Some(m) = metrics {
        m.files_parsed += 1;
    }
    fir.source = Some(content);
    fir.suppressed = suppressed;
    Ok(Some(fir))
}

pub fn parse_python_project(root: &Path) -> anyhow::Result<HashMap<String, FileIR>> {
    languages::python::parse_python_project(root)
}

pub fn parse_rust_project_uncached(root: &Path) -> anyhow::Result<HashMap<String, FileIR>> {
    languages::rust::parse_rust_project_uncached(root)
}

#[derive(Serialize, Deserialize)]
struct CachedFile {
    hash: String,
    module: String,
    ir: FileIR,
}

#[derive(Default, Serialize, Deserialize)]
struct CacheData {
    files: HashMap<String, CachedFile>,
}

pub fn parse_python_project_cached(
    root: &Path,
    cache_path: &Path,
    mut metrics: Option<&mut ParserMetrics>,
) -> anyhow::Result<(HashMap<String, FileIR>, usize)> {
    let mut cache: CacheData = fs::read_to_string(cache_path)
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or_default();

    let mut modules = HashMap::new();
    let mut parsed = 0usize;
    let mut stack = vec![root.to_path_buf()];
    let mut seen = HashSet::new();

    while let Some(dir) = stack.pop() {
        for entry in fs::read_dir(&dir)? {
            let entry = entry?;
            let path = entry.path();
            if path.is_dir() {
                stack.push(path);
            } else if path.extension().and_then(|e| e.to_str()) == Some("py") {
                let canonical = path.to_string_lossy().into_owned();
                seen.insert(canonical.clone());
                let content = fs::read_to_string(&path)?;
                let h = blake3::hash(content.as_bytes()).to_hex().to_string();
                if let Some(c) = cache.files.get(&canonical) {
                    if c.hash == h {
                        if let Some(m) = metrics.as_deref_mut() {
                            m.cache_hits += 1;
                        }
                        modules.insert(c.module.clone(), c.ir.clone());
                        continue;
                    }
                }
                let mut fir = FileIR::new(canonical.clone(), "python".into());
                if let Err(e) = languages::python::parse_python(&content, &mut fir) {
                    if let Some(m) = metrics.as_deref_mut() {
                        m.parse_errors += 1;
                    }
                    debug!(file = %canonical, error = ?e, "Failed to parse file");
                    continue;
                }
                if fir.symbol_types.contains_key("__parse_error__") {
                    if let Some(m) = metrics.as_deref_mut() {
                        m.parse_errors += 1;
                    }
                    debug!(file = %canonical, "File contains parse errors");
                    continue;
                }
                if let Some(m) = metrics.as_deref_mut() {
                    m.files_parsed += 1;
                }
                let rel = path.strip_prefix(root).unwrap_or(&path);
                let mut comps: Vec<String> = rel
                    .components()
                    .filter_map(|c| match c {
                        std::path::Component::Normal(os) => Some(os.to_string_lossy().into_owned()),
                        _ => None,
                    })
                    .collect();
                if comps.last().map(|s| s == "__init__.py").unwrap_or(false) {
                    comps.pop();
                }
                if let Some(last) = comps.last_mut() {
                    if last.ends_with(".py") {
                        *last = last.trim_end_matches(".py").to_string();
                    }
                }
                let module = comps.join(".");
                let names: Vec<String> = fir.symbols.keys().cloned().collect();
                for name in names {
                    if fir.symbols.get(&name).and_then(|s| s.def).is_some() {
                        fir.symbol_modules.insert(name, module.clone());
                    }
                }
                modules.insert(module.clone(), fir.clone());
                cache.files.insert(
                    canonical,
                    CachedFile {
                        hash: h,
                        module,
                        ir: fir,
                    },
                );
                parsed += 1;
            }
        }
    }
    cache.files.retain(|k, _| seen.contains(k));
    let cloned = modules.clone();
    for fir in modules.values_mut() {
        crate::languages::python::symbol_table::link_imports(fir, &cloned);
    }
    if let Ok(s) = serde_json::to_string(&cache) {
        let _ = fs::write(cache_path, s);
    }
    Ok((modules, parsed))
}

pub fn parse_rust_project(
    root: &Path,
    cache_path: &Path,
    metrics: Option<&mut ParserMetrics>,
) -> anyhow::Result<(HashMap<String, FileIR>, usize)> {
    languages::rust::parse_rust_project(root, cache_path, metrics)
}

pub fn parse_php_project(
    root: &Path,
    cache_path: &Path,
    metrics: Option<&mut ParserMetrics>,
) -> anyhow::Result<(HashMap<String, FileIR>, usize)> {
    languages::php::parse_php_project(root, cache_path, metrics)
}
