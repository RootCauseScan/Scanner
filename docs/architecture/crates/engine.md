---
id: engine
title: Engine
description: Analysis engine that evaluates rules on the intermediate representation
sidebar_position: 2
---

# engine

## Responsibility
Evaluates rules on the intermediate representation of files, handling parallel execution, caching and optional WASM support.

## Key functions and structures
- `analyze_files_cached`: analyzes files applying hash-based caching.
- `build_cfg`: builds control flow graphs for flow analysis.
- `Finding`: individual result with severity, location and message.

## Usage example
```rust
use engine::{analyze_files_cached, EngineConfig};
use loader::load_rules;
use parsers::parse_file;
use std::path::Path;

let rules = load_rules(Path::new("rules")).unwrap();
let file = parse_file(Path::new("a.yaml"), None, None).unwrap().unwrap();
let findings = analyze_files_cached(&[file], &rules, Path::new("cache.json"), &EngineConfig::default(), None);
assert!(findings.is_empty());
```
