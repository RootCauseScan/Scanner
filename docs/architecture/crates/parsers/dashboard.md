---
id: dashboard
title: Dashboard
description: Parsers to convert files from different languages to intermediate representation
sidebar_position: 1
---

# Dashboard

## Key functions and structures
- `detect_type`: determines the file type from its name or extension.
- `parse_file`: reads and transforms a file into `FileIR`.

## Maturity Guide

For detailed information about parser maturity levels (L1-L8), evaluation criteria and current status of each language, see the [Maturity Guide](./maturity.md).

## Supported languages

| Language | Level | Status |
| --- | --- | --- |
| Python | L5 | Intermediate |
| Rust | L2 | Syntactic |
| Dockerfile | L1 | Prototype |
| YAML | L1 | Prototype |
| HCL (Terraform) | L1 | Prototype |
| TypeScript | L1 | Prototype |
| JavaScript | L1 | Prototype |
| Go | L1 | Prototype |
| Java | L4 | Intermediate |
| PHP | L4 | Intermediate |
| Ruby | L1 | Prototype |

*For complete details on maturity levels and evaluation criteria, see the [Maturity Guide](./maturity.md).*

## Usage example
```rust
use parsers::parse_file;
use std::fs;
use std::path::Path;
let path = std::env::temp_dir().join("ex.yaml");
fs::write(&path, "a: 1").unwrap();
let ir = parse_file(Path::new(&path), None, None).unwrap().unwrap();
assert_eq!(ir.nodes[0].path, "a");
```

## Contributing new parsers

1. Create a module in `crates/parsers/src/languages/`.
2. Export it in `crates/parsers/src/languages/mod.rs` and update `detect_type` and `parse_file` in `crates/parsers/src/lib.rs`.
3. Add fixtures in `examples/fixtures/<language>` with `good` and `bad` cases.
4. Write tests in `crates/parsers/tests` using those fixtures.
