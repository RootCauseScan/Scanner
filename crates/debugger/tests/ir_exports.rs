use engine::{build_cfg, parse_file_with_events};
use parsers::build_dfg;
use std::path::Path;

fn load_fir() -> ir::FileIR {
    parse_file_with_events(
        Path::new("../../examples/fixtures/python/py.no-eval/bad.py"),
        None,
        None,
    )
    .unwrap()
    .unwrap()
}

#[test]
fn ast_to_dot() {
    let fir = load_fir();
    let ast = fir.ast.expect("ast");
    insta::assert_snapshot!(ast.to_dot());
}

#[test]
fn ast_to_json() {
    let fir = load_fir();
    let ast = fir.ast.expect("ast");
    insta::assert_snapshot!(ast.to_json().unwrap());
}

#[test]
fn cfg_to_dot() {
    let fir = load_fir();
    let cfg = build_cfg(&fir).expect("cfg");
    insta::assert_snapshot!(cfg.to_dot());
}

#[test]
fn cfg_to_json() {
    let fir = load_fir();
    let cfg = build_cfg(&fir).expect("cfg");
    insta::assert_snapshot!(cfg.to_json().unwrap());
}

#[test]
fn ssa_to_dot() {
    let mut fir = load_fir();
    build_dfg(&mut fir).unwrap();
    let dfg = fir.dfg.expect("dfg");
    insta::assert_snapshot!(dfg.to_dot());
}

#[test]
fn ssa_to_json() {
    let mut fir = load_fir();
    build_dfg(&mut fir).unwrap();
    let dfg = fir.dfg.expect("dfg");
    insta::assert_snapshot!(dfg.to_json().unwrap());
}
