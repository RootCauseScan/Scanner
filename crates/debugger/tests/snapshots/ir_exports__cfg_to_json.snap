---
source: crates/debugger/tests/ir_exports.rs
expression: cfg.to_json()
---
{
  "nodes": [
    {
      "id": 0,
      "line": 2,
      "code": "eval(expr)"
    }
  ],
  "edges": []
}
